pkg/cmd/browse/browse.go

```diff
@@ -0,0 +1,250 @@
+package browse
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/api"
+	"github.com/cli/cli/internal/ghrepo"
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
+
+type browser interface {
+	Browse(string) error
+}
+
+type BrowseOptions struct {
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+	BaseRepo   func() (ghrepo.Interface, error)
+	Browser    browser
+
+	SelectorArg   string
+	AdditionalArg string
+
+	ProjectsFlag bool
+	WikiFlag     bool
+	SettingsFlag bool
+	BranchFlag   bool
+	//LineFlag     bool
+}
+
+type exitCode int
+
+const (
+	exitSuccess      exitCode = 0
+	exitNotInRepo    exitCode = 1
+	exitTooManyFlags exitCode = 2
+	exitTooManyArgs  exitCode = 3
+	exitExpectedArg  exitCode = 4
+	exitInvalidCombo exitCode = 5
+)
+
+func NewCmdBrowse(f *cmdutil.Factory) *cobra.Command {
+
+	opts := &BrowseOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+		Browser:    f.Browser,
+		BaseRepo:   f.BaseRepo,
+	}
+
+	cmd := &cobra.Command{
+		Long:  "Work with GitHub in the browser",       // displays when you are on the help page of this command
+		Short: "Open GitHub in the browser",            // displays in the gh root help
+		Use:   "browse {<number> | <file> | <branch>}", // necessary!!! This is the cmd that gets passed on the prompt
+		Args:  cobra.RangeArgs(0, 2),                   // make sure only one arg at most is passed
+		Example: heredoc.Doc(`
+			$ gh browse
+			#=> Opens repository in browser
+
+			$ gh browse 217
+			#=> Opens issue or pull request 217
+
+			$ gh browse --settings
+			#=> Opens repository settings in browser
+
+			$ gh browse src/fileName:312 --branch main
+			#=> Opens src/fileName at line 312 in main branch
+		`),
+		Annotations: map[string]string{
+			"IsCore": "true",
+			"help:arguments": heredoc.Doc(`
+				A browse location can be supplied as an argument in any of the following formats:
+				- by number, e.g. "123"; or
+				- by file or branch name, e.g. "main.java" or "trunk".
+			`),
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+
+			if len(args) > 1 {
+				opts.AdditionalArg = args[1]
+			}
+
+			if len(args) > 0 {
+				opts.SelectorArg = args[0]
+			}
+			openInBrowser(cmd, opts) // run gets rid of the usage / runs function
```

Comment:
As I've already alluded to, you'll want to have your `openInBrowser` return just one value: an `error`. Once this is a `RunE` instead of a `Run`, you can then end this function with `return openInBrowser(cmd, opts)` to bubble up any returned `error` to `cmd/gh/main.go`.